libraries/AP_Scripting/examples/NAU7802.lua

```diff
@@ -0,0 +1,288 @@
+    return 1
+  end
+
+  if getBit(3, 2) then
+   return 2
+  end
+
+  -- Calibration passed
+  return 0;
+end
+
+-- Set the readings per second
+-- 10, 20, 40, 80, and 320 samples per second is available
+local function setSampleRate(rate)
+  if rate > 7 then
+    rate = 7
+  end
+
+  local value = i2c_bus:read_registers(2)
+  value = value & 143 -- Clear CRS bits
+  value = value | (rate << 4)  -- Mask in new CRS bits
+
+  return i2c_bus:write_register(2, value)
+end
+
+-- Select between 1 and 2
+local function setChannel(channelNumber)
+  if (channelNumber == 0) then
+    return clearBit(7, 2); -- Channel 1 (default)
+  else
+    return setBit(7, 2); -- Channel 2
+  end
+end
+
+-- Power up digital and analog sections of scale
+local function powerUp()
+  setBit(1, 0)
+  setBit(2, 0)
+
+  -- Wait for Power Up bit to be set - takes approximately 200us
+  local counter = 0;
+  while (true) do
+    if getBit(3, 0) then
+      break -- Good to go
+    end
+    if counter > 100 then
+      return false -- Error
+    end
+    counter = counter + 1
+  end
+  return true
+end
+
+-- Puts scale into low-power mode
+local function powerDown()
+  clearBit(1, 0)
+  return (clearBit(2, 0))
+end
+
+-- Resets all registers to Power Of Defaults
+local function reset()
+  setBit(0, 0) -- Set RR
+  return clearBit(0, 0) -- Clear RR to leave reset state
+end
+
+-- Set the onboard Low-Drop-Out voltage regulator to a given value
+-- 2.4, 2.7, 3.0, 3.3, 3.6, 3.9, 4.2, 4.5V are available
+local function setLDO(ldoValue)
+  if ldoValue > 7 then
+    ldoValue = 7 -- Error check
+  end
+
+  -- Set the value of the LDO
+  local value = i2c_bus:read_registers(1)
+  value = value & 199 -- Clear LDO bits
+  value = value | (ldoValue << 3) -- Mask in new LDO bits
+  i2c_bus:write_register(1, value);
+
+  return setBit(7, 0) -- Enable the internal LDO
+end
+
+-- Set the gain
+-- x1, 2, 4, 8, 16, 32, 64, 128 are avaialable
+local function setGain(gainValue)
+  if gainValue > 7 then
+    gainValue = 7 -- Error check
+  end
+
+  local value = i2c_bus:read_registers(1)
+  value = value & 248 -- Clear gain bits
+  value = value | gainValue -- Mask in new bits
+
+  return i2c_bus:write_register(1, value)
+end
+
+-- Get the revision code of this IC
+local function getRevisionCode()
+  revisionCode = i2c_bus:read_registers(31)
+  return revisionCode & 0x0F
+end
+
+-- Returns 24-bit reading
+-- Assumes CR Cycle Ready bit (ADC conversion complete) has been checked to be 1
+local function getReading()
+
+  local MSB = i2c_bus:read_registers(18)
+  local MID = i2c_bus:read_registers(19)
+  local LSB = i2c_bus:read_registers(20)
+
+  if MSB and MID and LSB then
+
+    local value = (MSB << 16) | (MID << 8) | LSB
+  -- Quite likely to be a signed vs unsigned error here !! need to check
+  
+  --  local sign = (value & (1 << 23)) > 0
+  --  if sign then
+  --    value = 16777215 ~ value -- 23 bit not
+  --   end
+  -- gcs:send_text(0,string.format("%u, %u, %u, %u",MSB,MID,LSB,value));
```

Comment:
```suggestion
    local sign = (value & (1 << 23)) > 0

    if sign then
        value = 16777215 ~ value -- 24 bit XOR to flip the bits for two's compliment
        value = (value * -1) - 1
    end
```

Reply:
You could probably string pack and unpack again to get this too. The suggestion is from the thrust test script.

Reply:
Applied this patch